jcat.c

```diff
@@ -0,0 +1,390 @@
+};
+
+static const char help_msg[] =
+	"jcat: netcat tool for j1939\n"
+	"Usage: jcat FROM TO\n"
+	" FROM / TO	- or [IFACE][:[SA][,[PGN][,NAME]]]\n"
+	"Options:\n"
+	" -i <infile>	(default stdin)\n"
+	" -s[=LEN]	Initial send of LEN bytes dummy data\n"
+	" -r		Receive data\n"
+	"\n"
+	"Example:\n"
+	"jcat -i some_file_to_send  can0:0x80 :0x90,0x12300\n"
+	"jcat can0:0x90 -r > /tmp/some_file_to_receive\n"
+	"\n"
+	;
+
+static const char optstring[] = "?i:vs::rp:cnw::";
+
+
+static void jcat_init_sockaddr_can(struct sockaddr_can *sac)
+{
+	sac->can_family = AF_CAN;
+	sac->can_addr.j1939.addr = J1939_NO_ADDR;
+	sac->can_addr.j1939.name = J1939_NO_NAME;
+	sac->can_addr.j1939.pgn = J1939_NO_PGN;
+}
+
+static void parse_canaddr(char *spec, struct sockaddr_can *paddr)
+{
+	char *str;
+
+	str = strsep(&spec, ":");
+	if (strlen(str))
+		paddr->can_ifindex = if_nametoindex(str);
+
+	str = strsep(&spec, ",");
+	if (str && strlen(str))
+		paddr->can_addr.j1939.addr = strtoul(str, NULL, 0);
+
+	str = strsep(&spec, ",");
+	if (str && strlen(str))
+		paddr->can_addr.j1939.pgn = strtoul(str, NULL, 0);
+
+	str = strsep(&spec, ",");
+	if (str && strlen(str))
+		paddr->can_addr.j1939.name = strtoul(str, NULL, 0);
+}
+
+static int jcat_sendfile(struct jcat_priv *priv, int out_fd, int in_fd,
+			 off_t *offset, size_t count)
+{
+	int ret = EXIT_SUCCESS;
+	off_t orig = 0;
+	char *buf;
+	size_t to_read, num_read, num_sent, tot_sent, buf_size;
+	int round = 0;
+
+	buf_size = min((size_t)J1939_MAX_ETP_PACKET_SIZE, count);
+	buf = malloc(buf_size);
+	if (!buf) {
+		warn("can't allocate buf");
+		ret = EXIT_FAILURE;
+		goto do_nofree;
+	}
+
+	if (offset != NULL) {
+
+		/* Save current file offset and set offset to value in '*offset' */
+
+		orig = lseek(in_fd, 0, SEEK_CUR);
+		if (orig == -1) {
+			ret = EXIT_FAILURE;
+			goto do_free;
+		}
+		if (lseek(in_fd, *offset, SEEK_SET) == -1) {
+			ret = EXIT_FAILURE;
+			goto do_free;
+		}
+	}
+
+	tot_sent = 0;
+
+	while (count > 0) {
+		to_read = min(buf_size, count);
+
+		num_read = read(in_fd, buf, to_read);
+		if (num_read == -1) {
+			ret = EXIT_FAILURE;
+			goto do_free;
+		}
+		if (num_read == 0)
+			break; /* EOF */
+
+		if (priv->valid_peername && !priv->todo_connect)
+			num_sent = sendto(out_fd, buf, num_read, 0,
+					(void *)&priv->peername,
+					sizeof(priv->peername));
+		else
+			num_sent = send(out_fd, buf, num_read, 0);
+
+		if (num_sent == -1) {
+			warn("sendfile: write() transferr error");
+			ret = EXIT_FAILURE;
+			goto do_free;
+		}
+
+		if (num_sent == 0) /* Should never happen */ {
+			warn("sendfile: write() transferred 0 bytes");
+			ret = EXIT_FAILURE;
+			goto do_free;
+		}
+
+		if (num_sent != num_read) {
+			warn("sendfile: write() not full transfer: %zi %zi",
+			     num_sent, num_read);
+			ret = EXIT_FAILURE;
+			goto do_free;
+		}
+
+		round++;
+		count -= num_sent;
+		tot_sent += num_sent;
+	}
+
+	if (offset != NULL) {
+		/* Return updated file offset in '*offset', and reset the file offset
```

Comment:
`if (!offset)`

Reply:
@olerem pointed out that should be `if (offset)`. doh!.
Fixed with https://github.com/linux-can/can-utils/pull/117